File: LeetCode/199_binary_tree_right_side_view/solve.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.que = []
        self.leaf_num = 0

    def rightSideView(self, root):
        ret = []
        if root:
            self.que.append(root)
            self.leaf_num = 0
            remain_leaf = 1
            while self.que:
                node = self.que.pop(0)
                tmp = node                      # override the right-most leaf value
                self.append_node(node.left)
                self.append_node(node.right)

                remain_leaf -= 1

                if remain_leaf == 0:            # current node is the right-most leaf in its depth level
                    ret.append(tmp.val)
                    remain_leaf = self.leaf_num
                    self.leaf_num = 0
        return ret

    def append_node(self, node):
        if node:
            self.que.append(node)
            self.leaf_num += 1                  # gather the info of next depth level

File: LeetCode/199_binary_tree_right_side_view/solve.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode *root) {
        vector<int> ret;
        queue<TreeNode *> que;
        if (root) {
            unsigned int leaf_num = 0, remain_leaf = 1;
            que.push(root); 

            while (!que.empty()) {
                TreeNode *node = que.front();
                que.pop();
                if (node->left) {
                    ++leaf_num;
                    que.push(node->left);
                }
                if (node->right) {
                    ++leaf_num;
                    que.push(node->right);
                }
                if (--remain_leaf == 0) {
                    ret.push_back(node->val);
                    remain_leaf = leaf_num;
                    leaf_num = 0;
                }
            }
        }
        return ret;
    }
};
